node-repository/src/main/java/com/yahoo/vespa/hosted/provision/provisioning/NodeIndices.java

```diff
@@ -0,0 +1,57 @@
+// Copyright Verizon Media. Licensed under the terms of the Apache 2.0 license. See LICENSE in the project root.
+package com.yahoo.vespa.hosted.provision.provisioning;
+
+import com.yahoo.config.provision.ClusterSpec;
+import com.yahoo.vespa.hosted.provision.NodeList;
+
+import java.util.List;
+
+import static java.util.Comparator.naturalOrder;
+
+/**
+ * Tracks indices of a node cluster, and proposes the index of the next allocation.
+ *
+ * @author jonmv
+ */
+class NodeIndices {
+
+    private final boolean compact;
+    private final List<Integer> used;
+
+    private int last;
+    private int probe;
+
+    NodeIndices(NodeList nodes, boolean compact) {
+        this.compact = compact;
+        this.used = nodes.mapToList(node -> node.allocation().get().membership().index());
+        this.last = compact ? -1 : used.stream().max(naturalOrder()).orElse(-1);
+        this.probe = last;
+    }
+
+    /** Returns the next available index and commits to using it. Throws if a probe is ongoing. */
+    int next() {
+        if (probe != last)
+            throw new IllegalStateException("Must commit ongoing probe before calling 'next'");
+
+        probeNext();
+        commitProbe();
+        return last;
+    }
+
+    /** Returns the next available index, without committing to using it. May be called multiple times. */
+    int probeNext() {
```

Comment:
Each successive invocation will increase probe by 1 which is not what we want?

Reply:
Yes, it's what we want. Successive probes should yield increasing values—a commit is for all of them at once.

Reply:
I guess the "May be called multiple times." could be more clear, then.

Reply:
Aha, ok. Yes fwiw my expectation upon reading that was it would return the same index each time.